ibis/expr/analysis.py

```diff
@@ -1265,3 +1265,67 @@ def _visit_table(self, expr):
                 and base_table.op() not in self.query_roots
             ):
                 self.foreign_table = expr
+
+
+@functools.singledispatch
+def _rewrite_filter(op, _, **kwargs):
+    raise NotImplementedError(type(op))
+
+
+@_rewrite_filter.register(ops.Reduction)
+def _rewrite_filter_reduction(_, expr, name: str | None = None):
+    # Find the table that this reduction references.
+
+    # TODO: what about reductions that reference a join that isn't visible
+    # at this level? Means we probably have the wrong design, but will have
+    # to revisit when it becomes a problem.
+    aggregation, _ = reduction_to_aggregation(expr, default_name=name)
+    return aggregation.to_array()
+
+
+@_rewrite_filter.register(ops.Any)
+@_rewrite_filter.register(ops.TableColumn)
+@_rewrite_filter.register(ops.Literal)
+@_rewrite_filter.register(ops.ExistsSubquery)
+@_rewrite_filter.register(ops.NotExistsSubquery)
+@_rewrite_filter.register(ops.Window)
+def _rewrite_filter_subqueries(_, expr, **kwargs):
+    return expr
+
+
+@_rewrite_filter.register(ops.Binary)
+def _rewrite_filter_binary(op, expr, **kwargs):
```

Comment:
Can you add comments describing what each overload does?

Reply:
Done, added docstrings.